tarantool/connection_pool.py

```diff
@@ -0,0 +1,406 @@
+        for addr in addrs:
+            ok, msg = validate_address(addr)
+            if not ok:
+                raise ConfigurationError(msg)
+        self.addrs = addrs
+
+        # Create connections
+        self.pool = {}
+        self.refresh_delay = refresh_delay
+        self.strategy = strategy_class(self.pool)
+
+        for addr in self.addrs:
+            key = self._make_key(addr)
+            self.pool[key] = PoolUnit(
+                addr=addr,
+                conn=Connection(
+                    host=addr['host'],
+                    port=addr['port'],
+                    user=user,
+                    password=password,
+                    socket_timeout=socket_timeout,
+                    reconnect_max_attempts=reconnect_max_attempts,
+                    reconnect_delay=reconnect_delay,
+                    connect_now=False, # Connect in ConnectionPool.connect()
+                    encoding=encoding,
+                    call_16=call_16,
+                    connection_timeout=connection_timeout)
+            )
+
+        if connect_now:
+            self.connect()
+
+    def __del__(self):
+        self.close()
+
+    def _make_key(self, addr):
+        return json.dumps(addr)
+
+    def _get_new_state(self, key, unit):
+        conn = unit.conn
+
+        if conn.is_closed():
+            try:
+                conn.connect()
+            except NetworkError as e:
+                msg = "Failed to connect to {0}:{1}".format(
+                    unit.addr['host'], unit.addr['port'])
+                warn(msg, ClusterConnectWarning)
+                return InstanceState(Status.UNHEALTHY)
+
+        try:
+            resp = conn.call('box.info')
+        except Exception as e:
+            msg = "Failed to get box.info for {0}:{1}, reason: {2}".format(
+                unit.addr['host'], unit.addr['port'], repr(e))
+            warn(msg, ClusterTolopogyWarning)
+            return InstanceState(Status.UNHEALTHY)
+
+        try:
+            ro = resp.data[0]['ro']
+        except Exception as e:
+            msg = "Incorrect box.info response from {0}:{1}".format(
+                unit.addr['host'], unit.addr['port'])
+            warn(msg, ClusterTolopogyWarning)
+            return InstanceState(Status.UNHEALTHY)
+
+        return InstanceState(Status.HEALTHY, ro)
+
+    def _refresh_state(self, key):
+        unit = self.pool[key]
+        conn = unit.conn
+
+        state = self._get_new_state(key, unit)
+        if state != unit.state:
+            unit.state = state
+            self.strategy.update()
+
+    def close(self):
+        for unit in self.pool.values():
+            unit.request_process_enabled = False
+            unit.thread.join()
+
+            if not unit.conn.is_closed():
+                unit.conn.close()
+
+    def is_closed(self):
+        return functools.reduce(
+            lambda x, y:
+                x.request_process_enabled or y.request_process_enabled,
+            self.pool)
```

Comment:
I have several thoughts about this method:
1. Do I understand it correctly that this method returns `True` if at least one pool unit have `request_process_enabled == True`? If that's so this method is actually `is_not_closed`.
2. It seems to me it can be implemented easier. Implementation with `any(unit.request_process_enabled for unit in self.pool.values())` would be equal to your logic If I got it correctly. If my remark in point 1 is correct, then this method can be implemented with `all(unit.request_process_enabled == False for unit in self.pool.values())`
3. Is there a test for this method?

Reply:
Thank you for each of your thoughts! Yes, I'm
* missing a test for the case,
* thus I missed that logic is invalid,
* and I haven't written in Python for a pretty long time so I forgot about `all` and `any`.

Fixed the logic, used `all` to implement (at first I thought about using `not` `any` -- it may be faster, but harder to read, and since it's not a performance critical place `all` suites better) and added the test.